content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,234 @@ Like a segment tree on values rather than indices.
+using namespace std;
+constexpr int MAX_VAL = 1e9 + 2;
+
+struct Segment {
+	Segment *left = nullptr, *right = nullptr;
+	int l, r, mid;
+	bool children = false;
+	vector<pair<int, int>> indices;  // Index, Value
+	vector<int> prefix_b;
+
+	Segment(int l, int r, const vector<pair<int, int>> &indices)
+	    : l(l), r(r), mid((r + l) / 2), indices(indices) {
+
+		// Calcultes the prefix B
+		int i = 1;
+		int j = 0;
+		prefix_b.resize(indices.size() + 1);
+		for (auto [ind, val] : indices) {
+			if (val < mid) j++;
+			prefix_b[i++] = j;
+		}
+	}
+
+	// Sparse since values can go up to 1e9
+	void update() {
+		if (children) { return; }
+		children = true;
+		if (r - l > 1) {
+			// Split the indices for left and right child
+			vector<pair<int, int>> leftIndices, rightIndices;
+			partition_copy(indices.begin(), indices.end(), leftIndices.begin(),
+			               rightIndices.begin(), [this](const pair<int, int> &elem) {
+				               return elem.second < mid;
+			               });
+
+			left = new Segment(l, mid, leftIndices);
+			right = new Segment(mid, r, rightIndices);
+		}
+	}
+
+	int find_k_smallest(int a, int b, int k) {
+		update();
+		if (r - l <= 1) { return l; }
+
+		int lb = prefix_b[a];
+		int lr = prefix_b[b];
+		int inLeft = lr - lb;  // Amount of values in range (a,b) that are less the mid
+
+		if (k <= inLeft) {
+			return left->find_k_smallest(lb, lr, k);  // Appears in left
+		} else {
+			return right->find_k_smallest(a - lb, b - lr,
+			                              k - inLeft);  // Appears in right
+		}
+	}
+};
+
+int main() {
+	int n, q;
+	cin >> n >> q;
+
+	vector<pair<int, int>> indices;
+	for (int i = 0; i < n; ++i) {
+		int v;
+		cin >> v;
+		indices.emplace_back(i, v);
+	}
+	Segment seg(0, MAX_VAL, indices);
+
+	for (int i = 0; i < q; ++i) {
+		int a, b, k;
+		cin >> a >> b >> k;
+		k++;
+		cout << seg.find_k_smallest(a, b, k) << " ";
+	}
+}
+```
+</CPPSection>
+</LanguageSection>
+
+
+
+## Supporting updates
+
+Let's support updates of type:
+	- change value at index $i$ to $y$
+
+We can traverse down to the leaf to remove the old element and also traverse down to add the new element.
+
+So what do the updates change?
+ -
+ Our indices vector
+ Our prefix vector
+
+To change the indices vector, what we can do is, instead of storing a vector, use a set.
+Then erasing and adding values becomes easy.
 
-<IncompleteSection />
+On the other hand, to change the prefix vector, since each update could change our prefix vector a lot, we can't maintain just the normal vector. What we could do is use a sparse segment tree.
+	- erasing and inserting can be done by just setting the value to 0 or 1 at the specific index
+	- querying for a prefix can be done by querying the segment tree from 0 to $i$
+This approach is not memory efficient and requires a segment tree's implementation.
+A more friendly approach would be using an order statistics tree.
+Such that querying for a prefix would be equivalent to order_of_key($i$)
```

Comment:
provide a snippet that shows how this would be implemented with order_of_key?

Reply:
Alright I have added it.

Reply:
also probably clarify that ost is exclusive to c++

Reply:
Alright I briefly explained about it